### PR TITLE
[topgen] Stop pretending it's not OpenTitan-specific

### PR DESCRIPTION
**Pull request by imphil**
Topgen was initially designed to be a rather generic toplevel generation
tool. That's not what it is today: it is heavily biased towards
generating OpenTitan toplevel designs -- which is exactly what we need
in OpenTitan. Instead of trying to keep things generic, this commit goes
the first step towards an "opinionated topgen".

As first step, all templates are stored not in the toplevel that is to
be generated, but next to topgen: other components in OpenTitan expect
files generated from these templates to be around exactly like that, so
we cannot simply modify them at will.

If necessary later we can add an override/extension mechanism of some
sort.

This commit tries to be rather minimal in the changes to topgen.py;
there's more to clean-up, but since we have at least three people
working on the same file at the moment I'm trying to keep the diff to a
minimum.

### CODE REVIEW

**Comment by imphil** [comment]
(Only the [topgen] commit is relevant, the other one is in #5475 and will be removed before merging; I only wanted to save myself a rebase.)

**Comment by tjaychen** [comment]
i see where you're going with this.  In the future, if we start building a drastically different looking top, where would be the right place to place those template files? (maybe completely different xbar configuration, however some of the hw/ip's are still reused).

To me, each top represents almost a ...certain "class" or design, so what would be the right way to handle a different class? Would it still be called topgen?

File: util/topgen.py
Comment (sriyerg) [inline util/topgen.py:1039]: Would it be useful to keep this (have it default to `util/topgen/templates`) for future 'tops' that may need different set of templates?

**Comment by imphil** [comment]
Thanks for having a look @sriyerg and @tjaychen!

To answer your questions at the same time: I realize that having configurability is great, and it's something I'd like to bring back -- once we know what exactly needs to be configured. We started topgen with the goal of producing generic toplevels, and over time added more and more things to make it less generic. That's fine, of course, our primary focus here is delivering OpenTitan, not a nice extensible framework for toplevel generation.

That being said, having some configurability is something I'm sure we'll need in the future, probably even reasonably soon. What I don't know yet, however, is what exactly we'll need to configure/extend/replace.

I'm therefore hoping we could go down this route:
* Acknowledge that *currently* topgen is rather specialized, as this PR does.
* Refactor some of the code we have in topgen currently with the goal to make it easier to maintain and think ahead of things which we'd like to swap out in the future. For example, move some of the `generate_xzy()` functions in `topgen.py` into reusable code pieces.
* Once we have a need for a very different toplevel design, we can then do more refactoring and provide the right extension points -- maybe template overrides, maybe the ability to provide custom `generate_plic()`-like functions, maybe something totally else.

Why not go the full way now, you might ask? I'd like to get us quickly to the point where we can (with less hacks) build different FPGA toplevels, and this PR already makes it slightly easier, as the removed files in top_englishbreakfast show. Going much beyond that will be a larger design task with a fair amount of discussion (and it should be!), and I don't feel we have the time for it right now.

Final note @tjaychen:
> To me, each top represents almost a ...certain "class" or design

It's funny you bring that up. Having talked to multiple people it seems everybody has a slightly different definition of what a "toplevel" is. Once we're past silver I'll try to summarize that and find agreement to ensure we actually think about the same thing when we use that word.

**Review by rswarbrick — APPROVED** [review]
I'm a big fan of this sort of thing. I haven't yet seen a "generic" framework with a single user that wasn't unusable when the second user came along. Let's refactor when the need arises.

**Comment by eunchan** [comment]
I think we should keep the template file in the top directory. Making it easier to modify (does not have fear to touch `util/` direcotry for digital designers) was the intention of placing the template to the top directory. And, it helped, I think.

While get down to the road to this point, we've modified the template file a lot. In certain cases, it was inevitable as topgen does not support some features (module inter-connection, for example). And still some module is not fit into topgen yet. No general way to placing memory. 

Moving the template to topgen directory, IMO, possibly make the template stagnated. 

Just my 2¢

**Comment by tjaychen** [comment]
definitely agree with @imphil that after silver we should collectively try to define what "toplevel" is.  To me this ties a lot into the FPGA discussion as well, and I think in general there are several different opinions on how this should work. 

I also agree that topgen will end up being very top specific.. maybe there are pieces that can be re-used by other projects, but that would probably take a lot more care and refactoring.  

As you're looking at this though, could we also give some thought to @eunchan proposal [here](https://docs.google.com/document/d/1jimqHdR1x8on3jwzsi8SHTdhXFvujHj-EFk-eGU9mX4/edit?usp=sharing)?  One part that always bothered me is the `hw/ip` module templating, and the fact that local generation scripts inevitably get out of sync vs the top level.  Even if `topgen` itself is specific to a top, it feels like we should at least try to keep the `ip` versions somewhat self contained. 

To Eunchan's point about keeping the templates near the design files, I don't personally have a strong opinion either way, but I feel like it would be best to be consistent.  Right now most of the templates are kept near `hw/ip`, with the reg / xbar templates kept near `util`.  I think if we're going to pick one or the other, would it make sense to align everyone?

**Comment by imphil** [comment]
Thanks @tjaychen for bringing the "IPGEN" proposal up. I wasn't aware of that and yes, that's exactly the same problem I'm trying to solve right now as well. I'll read through it and get back.

**Comment by imphil** [comment]
> As you're looking at this though, could we also give some thought to @eunchan proposal here?

I have now a first draft of the ipgen tool working and will push it into a PR after some cleanup tomorrow. It already helps to make topgen a bit friendlier.

However, the more I look at topgen, the more it is clear how tied it is to OpenTitan as we have it today (and top_earlgrey), and there is simply no well-defined API between topgen, the consolidated top hjson (which is passed to the template), and the templates itself. I hope we'll untangle all of that in the future (and I'll be working towards that as well), but in the meantime, I'd really love to get the templates out of top_earlgrey and simplify the code along with it, to unblock work on getting multiple FPGAs working.

@eunchan @tjaychen would you be OK with me merging this change, and we can then work on improving the situation to provide a clear API and really exchangeable templates in the future?

**Comment by tjaychen** [comment]
> > As you're looking at this though, could we also give some thought to @eunchan proposal here?
> 
> I have now a first draft of the ipgen tool working and will push it into a PR after some cleanup tomorrow. It already helps to make topgen a bit friendlier.
> 
> However, the more I look at topgen, the more it is clear how tied it is to OpenTitan as we have it today (and top_earlgrey), and there is simply no well-defined API between topgen, the consolidated top hjson (which is passed to the template), and the templates itself. I hope we'll untangle all of that in the future (and I'll be working towards that as well), but in the meantime, I'd really love to get the templates out of top_earlgrey and simplify the code along with it, to unblock work on getting multiple FPGAs working.
> 
> @eunchan @tjaychen would you be OK with me merging this change, and we can then work on improving the situation to provide a clear API and really exchangeable templates in the future?

sounds good to me.

**Comment by eunchan** [comment]
OK!

**Pull request merged**
